core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala

```diff
@@ -2077,6 +2086,98 @@ class TaskSetManagerSuite
     assert(manager.resourceOffer("exec1", "host1", ANY)._1.isEmpty)
   }
 
+  test("SPARK-21040: Tasks are not speculated on decommissioning if speculation is disabled") {
```

Comment:
Do we really need this? This indeed the only behavior before SPARK-21040. And SPARK-21040 added a flag to allow users to switch to the new behavior.

This test just makes me feel that we're testing whether the flag works well. If we really need this test, I'd prefer to combine this test and the above one to reuse the codes instead of bringing too many duplicate codes to test the tiny difference.

Reply:
I think that this test is testing the right things but perhaps this is not the right place for them. Basically as you observed, it is merely testing that the flag is working as expected: In effect it is testing that `TaskSetManager#checkSpeculatableTasks` is returning false when speculation is disabled, which is quite trivial.

The other thing that it is checking (which is actually not covered very well directly) is that when a decommissioned executor is finally lost, its tasks are rerun and not anytime before that. I think while this is tested sort of at a higher level (in the worker unit tests), it is not directly validated in either the task set manager nor the task scheduler tests.

I hear you and I think I wasn't thinking very hard when I added the test. I am going to split it up into two pieces (and thus also reduce the code duplication):
- I will make this test just ensure that `TaskSetManager#checkSpeculatableTasks` returns false and that nothing is speculated when the speculation is disabled. It does not need to bother with decommissioning as such. 
- I will fold in the logic for ensuring that the task is rerun on executor lost in one of the task scheduler's decommissioning tests. 

Btw, on code duplication in tests: I think it should be okay since we really want each test to be independant to understand. Putting a lot of helper method, just makes it hard to understand. I think test code should be more biased towards independence/flexibility instead of repetition, unlike production code. But I can see it your way as well :-) 

Thanks for flagging this @Ngone51 and it helps improve the test code quality !.

Also cc' @holdenk for her thoughts.

Reply:
Thank you @agrawaldevesh for your detailed explanation.  

> In effect it is testing that TaskSetManager#checkSpeculatableTasks is returning false when speculation is disabled

IIUC, `TaskSetManager#checkSpeculatableTasks` won't be called when speculation is disabled: 

https://github.com/apache/spark/blob/8b4862953a879a9b3ba6f57e669efc383df68b7c/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala#L212-L217

So I think the test might be testing an impossible case.

> when a decommissioned executor is finally lost, its tasks are rerun and not anytime before that.

I don't have a strong opinion about this. It's good in terms of integration test. But I'd like to say that I think this's actually the original behavior of executor lose and we should already have some tests cover it. e.g., an executor has been already lost but the driver only gets to know it when the heartbeat timeout. During this time,  `TaskSetManager` also doesn't rerun tasks. IIUC, decommission is built on top of this kind of basis, so it ideally shouldn't change the original behavior indeed.

Reply:
You are right about the first part: `TaskSetManager#checkSpeculatableTasks` is only ever called directly (without TaskSchedulerImpl) in the `TaskSetManagerSuite`, and in all of those calls, speculation is indeed always enabled. 

So I think I can delete the test in the `TaskSetManagerSuite` and the corresponding change to `TaskSetManager#checkSpeculatableTasks` to check the `speculationEnabled` flag. 

But I would like to retain the test in `TaskSchedulerImplSuite`: I hear that it may be a redundant test but the newly added test is really quick and IMHO it never hurts to ensure invariants without relying on a lower level test -- ie the double testing does no harm and it provides some coverage even if the underlying behavior changes. So unless you have strong feelings here, how about I keep the additional test in `TaskSchedulerImplSuite` ?

Thanks !

Reply:
Sounds ok to me!

Reply:
Thanks, so if you have no other comments or blockers, would you be willing to mark your acceptance of the PR by approving it :-P ? Thanks,

Reply:
Yeah I'm ok with code duplication in tests. I tend to write helper functions to reduce this when I can but it's a personal choice. I think reverting this test or moving this test is fine though.

Reply:
I have already done the changes as stated in the comment above: This test has been effectively "moved" : The bulk of its logic as for decommissioning has been moved to `TaskSchedulerImplSuite` and the original in `TaskManagerSuite` has been removed. The test in `TaskSchedulerImplSuite` is very decom specific and does not test anything about speculation as such.